pex/specifier_sets.py

```diff
@@ -0,0 +1,331 @@
+        excluded_range = Range(lower=self.lower, upper=self.upper)
+        return other in excluded_range or excluded_range in other
+
+    def __str__(self):
+        return ",".join(map(str, (self.lower, self.upper)))
+
+
+@attr.s(frozen=True)
+class Range(object):
+    lower = attr.ib(default=None)  # type: Optional[LowerBound]
+    upper = attr.ib(default=None)  # type: Optional[UpperBound]
+    excludes = attr.ib(default=())  # type: Tuple[ExcludedRange, ...]
+
+    def __contains__(self, other):
+        # type: (Any) -> bool
+        if Range is not type(other):
+            return NotImplemented
+        return self.contains(other)
+
+    def contains(self, other):
+        # type: (Range) -> bool
+
+        if self.lower:
+            if not other.lower:
+                return False
+            elif self.lower.version > other.lower.version:
+                return False
+            elif (
+                self.lower.version == other.lower.version
+                and not self.lower.inclusive
+                and other.lower.inclusive
+            ):
+                return False
+
+        if self.upper:
+            if not other.upper:
+                return False
+            elif self.upper.version < other.upper.version:
+                return False
+            elif (
+                self.upper.version == other.upper.version
+                and not self.upper.inclusive
+                and other.upper.inclusive
+            ):
+                return False
+
+        for excluded_range in self.excludes:
+            if excluded_range.intersects(other):
+                return False
+
+        return True
+
+    def __str__(self):
+        # type: () -> str
+        parts = []
+        if self.lower:
+            parts.append(str(self.lower))
+        for excluded_range in self.excludes:
+            parts.append("!({})".format(excluded_range))
+        if self.upper:
+            parts.append(str(self.upper))
+        return ",".join(parts)
+
+
+def _increment_release(
+    version,  # type: Version
+    index,  # type: int
+):
+    # type: (...) -> Version
+
+    parsed_version = version.parsed_version
+
+    parts = []  # type: List[str]
+    if parsed_version.epoch != 0:
+        parts.append(str(parsed_version.epoch))
+        parts.append("!")
+    if parsed_version.release:
+        ceiling_release = list(parsed_version.release[:index])
+        ceiling_release.append(parsed_version.release[index] + 1)
+        parts.append(".".join(str(x) for x in ceiling_release))
+    if parsed_version.pre is not None:
+        parts.append("".join(str(x) for x in parsed_version.pre))
+    if parsed_version.post is not None:
+        parts.append(".post")
+        parts.append(str(parsed_version.post))
+    if parsed_version.dev is not None:
+        parts.append(".dev")
+        parts.append(str(parsed_version.dev))
+    if parsed_version.local is not None:
+        parts.append("+")
+        parts.append(parsed_version.local)
```

Comment:
At least for the compatibility operator `~=`, the suffix is ignored so I imagine it would be more correct to drop these for the ceiling version. I wonder if the same is not true also for the other operators?
e.g. from the packaging version specifiers docs example, the ceiling equiv. version drops these:
```
~= 2.2.post3
>= 2.2.post3, == 2.*

~= 1.4.5a4
>= 1.4.5a4, == 1.4.*
```

Reply:
Thanks - that's definitely a bug.